crates/rpc/src/gas_price.rs

```diff
@@ -1,112 +1,90 @@
+    pub async fn get(&self) -> Option<U256> {
+        if let Some(Value { gas_price, updated }) = *self.value.read().await {
+            if updated.elapsed() <= self.horizon {
+                tracing::debug!(?gas_price, from=?updated, "Using cached gas price value");
+                return Some(gas_price);
             }
+        }
 
-            // clear stale since it's not going to be useful for anyone
-            g.latest = None;
-
-            // this is an adaptation of https://fasterthanli.me/articles/request-coalescing-in-async-rust
-
-            if let Some(tx) = g.next.upgrade() {
-                // there's already an existing request being fulfilled
-                tx.subscribe()
-            } else {
-                let (tx, rx) = tokio::sync::broadcast::channel(1);
-
-                // the use of Weak works here, because the only strong reference is being sent to
-                // the async task, which upon completion holds the lock again while sending
-                // everyone listening the response, and clears the weak.
-                let tx = Arc::new(tx);
-
-                let inner = self.inner.clone();
-                let gateway = self.gateway.clone();
-
-                g.next = Arc::downgrade(&tx);
-
-                // Update the gas price from the starknet pending block.
-                tokio::spawn(async move {
-                    use starknet_gateway_types::reply::MaybePendingBlock;
-                    let gas_price = match gateway
-                        // Don't indefinitely retry as this could block the RPC request.
-                        .block_without_retry(pathfinder_common::BlockId::Pending)
-                        .await
-                    {
-                        Ok(b) => match b {
-                            MaybePendingBlock::Pending(b) => b.gas_price,
-                            MaybePendingBlock::Block(b) => match b.gas_price {
-                                Some(g) => g,
-                                None => {
-                                    tracing::debug!("Gas price missing in block");
-                                    let _ = tx.send(None);
-                                    return;
-                                }
-                            },
-                        },
-                        Err(reason) => {
-                            tracing::debug!(%reason, "Failed to fetch gas price");
-                            let _ = tx.send(None);
-                            return;
-                        }
-                    };
+        let mut guard = self.value.write().await;
 
-                    let now = std::time::Instant::now();
+        // Double-check for cached gas price, to avoid concurrent requests to the gateway.
+        if let Some(Value { gas_price, updated }) = *guard {
+            if updated.elapsed() <= self.horizon {
+                tracing::debug!(?gas_price, from=?updated, "Using cached gas price value");
```

Comment:
Can we maybe make this a `trace!()`-level message? These _could_ be useful when diagnosing an error but otherwise would mean we get a lot of messages when running with RUST_LOG=debug during normal operation.

The same goes for all non-error-related messages below.

Reply:
+1